In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('combinado.csv')
df.head()  # muestra las primeras filas del DataFrame

df_copy = df.copy()

In [3]:
df.describe()  # proporciona estadísticas descriptivas para las columnas numéricas

CODIGO DISTRITO    DEPARTAMENTO MUNICIPIO  \
count            9621     9397            9621      9621   
unique           8425      603              20       304   
top     18-01-2733-46   18-008  CIUDAD CAPITAL    ZONA 1   
freq                3      384            1536       628   

                                      ESTABLECIMIENTO           DIRECCION  \
count                                            9621                9576   
unique                                           4821                5452   
top     INSTITUTO NACIONAL DE EDUCACION DIVERSIFICADA  CABECERA MUNICIPAL   
freq                                              356                 252   

        TELEFONO                                      SUPERVISOR DIRECTOR  \
count       9055                                            9396     8233   
unique      4875                                             576     4433   
top     79480009  ELENA ELIZABETH SUCHITE GARNICA DE QUINTANILLA      ---   
freq          42                                             384      204   

                NIVEL   SECTOR    AREA   STATUS   MODALIDAD JORNADA  \
count            9621     9621    9621     9621        9621    9621   
unique              1        4       3        4           2       6   
top     DIVERSIFICADO  PRIVADO  URBANA  ABIERTA  MONOLINGUE   DOBLE   
freq             9621     8274    7738     6877        9379    3239   

                   PLAN DEPARTAMENTAL  
count              9621          9621  
unique               13            22  
top     DIARIO(REGULAR)        IZABAL  
freq               5866          1080

In [4]:

df.info()  # muestra información sobre el DataFrame, incluyendo los tipos de datos y la cantidad de valores no nulos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9621 entries, 0 to 9620
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   CODIGO           9621 non-null   object
 1   DISTRITO         9397 non-null   object
 2   DEPARTAMENTO     9621 non-null   object
 3   MUNICIPIO        9621 non-null   object
 4   ESTABLECIMIENTO  9621 non-null   object
 5   DIRECCION        9576 non-null   object
 6   TELEFONO         9055 non-null   object
 7   SUPERVISOR       9396 non-null   object
 8   DIRECTOR         8233 non-null   object
 9   NIVEL            9621 non-null   object
 10  SECTOR           9621 non-null   object
 11  AREA             9621 non-null   object
 12  STATUS           9621 non-null   object
 13  MODALIDAD        9621 non-null   object
 14  JORNADA          9621 non-null   object
 15  PLAN             9621 non-null   object
 16  DEPARTAMENTAL    9621 non-null   object
dtypes: object(17)
memory usage: 1.2+ 

In [5]:
df.iloc[:,1:].describe()

DISTRITO    DEPARTAMENTO MUNICIPIO  \
count      9397            9621      9621   
unique      603              20       304   
top      18-008  CIUDAD CAPITAL    ZONA 1   
freq        384            1536       628   

                                      ESTABLECIMIENTO           DIRECCION  \
count                                            9621                9576   
unique                                           4821                5452   
top     INSTITUTO NACIONAL DE EDUCACION DIVERSIFICADA  CABECERA MUNICIPAL   
freq                                              356                 252   

        TELEFONO                                      SUPERVISOR DIRECTOR  \
count       9055                                            9396     8233   
unique      4875                                             576     4433   
top     79480009  ELENA ELIZABETH SUCHITE GARNICA DE QUINTANILLA      ---   
freq          42                                             384      204   

                NIVEL   SECTOR    AREA   STATUS   MODALIDAD JORNADA  \
count            9621     9621    9621     9621        9621    9621   
unique              1        4       3        4           2       6   
top     DIVERSIFICADO  PRIVADO  URBANA  ABIERTA  MONOLINGUE   DOBLE   
freq             9621     8274    7738     6877        9379    3239   

                   PLAN DEPARTAMENTAL  
count              9621          9621  
unique               13            22  
top     DIARIO(REGULAR)        IZABAL  
freq               5866          1080

In [6]:
df.shape

(9621, 17)

### Describa el estado de los datos y las operaciones de limpieza que considera que hará

La mayoria de los datos estan completos. En algunas columnas no hay datos o hay simbolos que no estan en UTF-8 lo cual tendremos que arreglar para hacer una buena lectura de los datos. No consideramos necesario borrar ninguna fila por ahora pero si hay columnas que consideramos que no se necesitaran (Direccion, telefono, director)

#### Columnas:

Columnas:
- Codigo: Dropeada
- Telefono: Dropeada
- Director: Dropeada
- DISTRITO: Arreglar null values
- DEPARTAMENTO: Codificar
- MUNICIPIO: Codificar
- ESTABLECIMIENTO: Codificar
- DIRECCION: Arreglar null values, Codificar
- NIVEL: Codificar
- SECTOR: Codificar
- AREA: Codificar
- STATUS: Codificar
- MODALIDAD:Codificar
- JORNADA: Codificar
- PLAN: Codificar
- DEPARTAMENTAL: Codificar

In [7]:
# Find columns with null values
null_columns = df.isnull().sum()

# Display the result
print(null_columns)

CODIGO                0
DISTRITO            224
DEPARTAMENTO          0
MUNICIPIO             0
ESTABLECIMIENTO       0
DIRECCION            45
TELEFONO            566
SUPERVISOR          225
DIRECTOR           1388
NIVEL                 0
SECTOR                0
AREA                  0
STATUS                0
MODALIDAD             0
JORNADA               0
PLAN                  0
DEPARTAMENTAL         0
dtype: int64


Al usar los métodos isnull y sum del dataframe, podemos contar los respectivos valores nulos de cada columna del Dataframe. Las columnas con valores nulos son: Distrito, Dirección, teléfono, supervisor y director. Tales columnas son las que requieren tratamientos por nuestra parte. 

In [8]:
df.drop('CODIGO', axis=1, inplace=True)
df.drop('TELEFONO', axis=1, inplace=True)
df.drop('DIRECTOR', axis=1, inplace=True)
df.drop('SUPERVISOR', axis=1, inplace=True)

Usamos el método drop para eliminar las columnas Código, teléfono, director y supervisor; al tener el parámetro inplace= True, modificamos el dataframe original eliminando del mismo dichas columnas.

In [9]:
# Eliminar filas duplicadas en todo el DataFrame
df.drop_duplicates(inplace=True)

El método drop_duplicates nos permite eliminar columnas que contengan los mismos valores, i.e., que sean columnas duplicadas.

In [10]:
# Fill null values in the 'column_name' column with the previous value in the column
df['DIRECCION'].fillna(method='ffill', inplace=True)
# Fill null values in the 'column_name' column with the previous value in the column
df['DIRECCION'].fillna(method='ffill', inplace=True)
# Fill null values in the 'column_name' column with the previous value in the column
df['DISTRITO'].fillna(method='ffill', inplace=True)

Utilizamos el método fillna para reemplazar los datos NaN del dataframe. El método que utilizamos para dicho reemplazo es el fill forward; este método sustituye el valor faltante por el valor de la última observación disponible.

In [11]:
# Codificar variables categóricas utilizando one-hot encoding
df_encoded = pd.get_dummies(df, columns=['DEPARTAMENTO', 'MUNICIPIO', 'ESTABLECIMIENTO', 'DIRECCION', 'NIVEL', 'SECTOR', 'AREA', 'STATUS', 'MODALIDAD', 'JORNADA', 'PLAN', 'DEPARTAMENTAL'])

Utilizamos el método get_dummies para convertir las columnas de variables categóricas a columnas con variables dummies -con valores 1,0-.

In [12]:
df

DISTRITO  DEPARTAMENTO  MUNICIPIO  \
0      16-031  ALTA VERAPAZ      COBAN   
1      16-031  ALTA VERAPAZ      COBAN   
2      16-031  ALTA VERAPAZ      COBAN   
3      16-005  ALTA VERAPAZ      COBAN   
4      16-005  ALTA VERAPAZ      COBAN   
...       ...           ...        ...   
9616   19-021        ZACAPA   LA UNION   
9617   19-021        ZACAPA   LA UNION   
9618   19-015        ZACAPA      HUITE   
9619   19-015        ZACAPA      HUITE   
9620   19-020        ZACAPA  SAN JORGE   

                                        ESTABLECIMIENTO  \
0                                         COLEGIO COBAN   
1                      COLEGIO PARTICULAR MIXTO VERAPAZ   
2                               COLEGIO "LA INMACULADA"   
3              ESCUELA NACIONAL DE CIENCIAS COMERCIALES   
4     INSTITUTO NORMAL MIXTO DEL NORTE 'EMILIO ROSAL...   
...                                                 ...   
9616                     LICEO PARTICULAR MIXTO "JIREH"   
9617                    LICEO PARTICULAR MIXTO " JIREH"   
9618                            INSTITUTO DIVERSIFICADO   
9619  INSTITUTO MIXTO DE EDUCACION DIVERSIFICADA POR...   
9620  INSTITUTO MIXTO DE EDUCACIÓN DIVERSIFICADA POR...   

                                   DIRECCION          NIVEL       SECTOR  \
0     KM.2 SALIDA A SAN JUAN CHAMELCO ZONA 8  DIVERSIFICADO      PRIVADO   
1               KM 209.5 ENTRADA A LA CIUDAD  DIVERSIFICADO      PRIVADO   
2                  7A. AVENIDA 11-109 ZONA 6  DIVERSIFICADO      PRIVADO   
3                      2A CALLE 11-10 ZONA 2  DIVERSIFICADO      OFICIAL   
4                        3A AVE 6-23 ZONA 11  DIVERSIFICADO      OFICIAL   
...                                      ...            ...          ...   
9616                            BARRIO NUEVO  DIVERSIFICADO      PRIVADO   
9617                            BARRIO NUEVO  DIVERSIFICADO      PRIVADO   
9618                     BARRIO BUENOS AIRES  DIVERSIFICADO      OFICIAL   
9619                         BARRIO EL CAMPO  DIVERSIFICADO  COOPERATIVA   
9620                        BARRIO EL CENTRO  DIVERSIFICADO  COOPERATIVA   

        AREA   STATUS   MODALIDAD      JORNADA  \
0     URBANA  ABIERTA  MONOLINGUE     MATUTINA   
1     URBANA  ABIERTA  MONOLINGUE     MATUTINA   
2     URBANA  ABIERTA  MONOLINGUE     MATUTINA   
3     URBANA  ABIERTA  MONOLINGUE     MATUTINA   
4     URBANA  ABIERTA    BILINGUE   VESPERTINA   
...      ...      ...         ...          ...   
9616  URBANA  ABIERTA  MONOLINGUE     MATUTINA   
9617  URBANA  ABIERTA  MONOLINGUE  SIN JORNADA   
9618  URBANA  ABIERTA  MONOLINGUE     NOCTURNA   
9619  URBANA  ABIERTA  MONOLINGUE   VESPERTINA   
9620  URBANA  ABIERTA  MONOLINGUE     MATUTINA   

                                     PLAN DEPARTAMENTAL  
0                         DIARIO(REGULAR)  ALTA VERAPAZ  
1                         DIARIO(REGULAR)  ALTA VERAPAZ  
2                         DIARIO(REGULAR)  ALTA VERAPAZ  
3                         DIARIO(REGULAR)  ALTA VERAPAZ  
4                         DIARIO(REGULAR)  ALTA VERAPAZ  
...                                   ...           ...  
9616                      DIARIO(REGULAR)        ZACAPA  
9617  SEMIPRESENCIAL (UN DÍA A LA SEMANA)        ZACAPA  
9618                      DIARIO(REGULAR)        ZACAPA  
9619                      DIARIO(REGULAR)        ZACAPA  
9620                      DIARIO(REGULAR)        ZACAPA  

[8087 rows x 13 columns]

In [13]:
# vemos los valores unicos en cada columna del dataframe con el metodo .unique()
# guardamos en un diccionario. La llave es el nombre de la columna y la lista de valores unicos el el valor corresp
unique_values_dict = {}

# Loop through the columns and populate the dictionary
for column_name in df.columns:
    #print(len(df[column_name].unique()))
    unique_values_dict[column_name] = df[column_name].unique()
    #print("\n")

#print(unique_values_dict)

# lista de listas con los valores que queremos codificar
#list_of_lists = [unique_values_dict[column_name] for column_name in df.columns[df.columns.get_loc("SECTOR"):]]
list_of_lists = [unique_values_dict[column_name] for column_name in df.columns[df.columns.get_loc("SECTOR"):df.columns.get_loc(df.columns[-1])]]

#print(df.columns.get_loc("SECTOR"))

#list_of_lists

In [14]:
# vemos cuantos valores unicos existen por columna del dataframe
for key in unique_values_dict:
    print(key + " " + str(len(unique_values_dict[key])))

DISTRITO 603
DEPARTAMENTO 20
MUNICIPIO 304
ESTABLECIMIENTO 4821
DIRECCION 5452
NIVEL 1
SECTOR 4
AREA 3
STATUS 4
MODALIDAD 2
JORNADA 6
PLAN 13
DEPARTAMENTAL 22


Al observar los valores unicos por columna del dataframe, vemos que todas las escuelas tienen nivel diversificado, por lo tanto podemos eliminar dicha columna del dataframe. Por lo otro lado, vamos a codificar el sector, area, status, modalidad, jornada y plan.

La manera de codificar es la siguiente; cada columna tiene una lista de valores únicos y vamos a colocar como código la posición del dato observado, la respectiva posicion que ocupa dicho dato en la lista de valores unicos.

In [15]:
# hacemos la codificacion con el area
unique_values_area = unique_values_dict["SECTOR"]
# Create a mapping of values to positions in the unique values list
value_to_position = {value: index for index, value in enumerate(unique_values_area)}

df['encoded_SECTOR'] = df['SECTOR'].map(value_to_position)

# eliminamos la columna area despues de codificarla
df.drop('SECTOR', axis=1, inplace=True)

#df

In [16]:
# hacemos la codificacion con el area
unique_values_area = unique_values_dict["AREA"]
# Create a mapping of values to positions in the unique values list
value_to_position = {value: index for index, value in enumerate(unique_values_area)}

df['encoded_AREA'] = df['AREA'].map(value_to_position)

# eliminamos la columna area despues de codificarla
df.drop('AREA', axis=1, inplace=True)

#df

In [17]:
# hacemos la codificacion con el status
unique_values_area = unique_values_dict["STATUS"]
# Create a mapping of values to positions in the unique values list
value_to_position = {value: index for index, value in enumerate(unique_values_area)}

df['encoded_STATUS'] = df['STATUS'].map(value_to_position)

# eliminamos la columna area despues de codificarla
df.drop('STATUS', axis=1, inplace=True)

#df

In [18]:
# hacemos la codificacion con la modalidad
unique_values_area = unique_values_dict["MODALIDAD"]
# Create a mapping of values to positions in the unique values list
value_to_position = {value: index for index, value in enumerate(unique_values_area)}

df['encoded_MODALIDAD'] = df['MODALIDAD'].map(value_to_position)

# eliminamos la columna area despues de codificarla
df.drop('MODALIDAD', axis=1, inplace=True)

#df

In [19]:
# hacemos la codificacion con la jornada
unique_values_area = unique_values_dict["JORNADA"]
# Create a mapping of values to positions in the unique values list
value_to_position = {value: index for index, value in enumerate(unique_values_area)}

df['encoded_JORNADA'] = df['JORNADA'].map(value_to_position)

# eliminamos la columna area despues de codificarla
df.drop('JORNADA', axis=1, inplace=True)

In [20]:
# hacemos la codificacion con el PLAN
unique_values_area = unique_values_dict["PLAN"]
# Create a mapping of values to positions in the unique values list
value_to_position = {value: index for index, value in enumerate(unique_values_area)}

df['encoded_PLAN'] = df['PLAN'].map(value_to_position)

# eliminamos la columna area despues de codificarla
df.drop('PLAN', axis=1, inplace=True)

# puesto que todos los colegios tiene el nivel diversificado, eliminamos dicha columna
df.drop('NIVEL', axis=1, inplace=True)

# puesto que la columna departamento ya incluye esta informacion, eliminamos dicha columna
df.drop('DEPARTAMENTAL', axis=1, inplace=True)

# considerando que la direccion del colegio no es necesario para el analisis, eliminamos dicha columna
df.drop('DIRECCION', axis=1, inplace=True)

df

DISTRITO  DEPARTAMENTO  MUNICIPIO  \
0      16-031  ALTA VERAPAZ      COBAN   
1      16-031  ALTA VERAPAZ      COBAN   
2      16-031  ALTA VERAPAZ      COBAN   
3      16-005  ALTA VERAPAZ      COBAN   
4      16-005  ALTA VERAPAZ      COBAN   
...       ...           ...        ...   
9616   19-021        ZACAPA   LA UNION   
9617   19-021        ZACAPA   LA UNION   
9618   19-015        ZACAPA      HUITE   
9619   19-015        ZACAPA      HUITE   
9620   19-020        ZACAPA  SAN JORGE   

                                        ESTABLECIMIENTO  encoded_SECTOR  \
0                                         COLEGIO COBAN               0   
1                      COLEGIO PARTICULAR MIXTO VERAPAZ               0   
2                               COLEGIO "LA INMACULADA"               0   
3              ESCUELA NACIONAL DE CIENCIAS COMERCIALES               1   
4     INSTITUTO NORMAL MIXTO DEL NORTE 'EMILIO ROSAL...               1   
...                                                 ...             ...   
9616                     LICEO PARTICULAR MIXTO "JIREH"               0   
9617                    LICEO PARTICULAR MIXTO " JIREH"               0   
9618                            INSTITUTO DIVERSIFICADO               1   
9619  INSTITUTO MIXTO DE EDUCACION DIVERSIFICADA POR...               3   
9620  INSTITUTO MIXTO DE EDUCACIÓN DIVERSIFICADA POR...               3   

      encoded_AREA  encoded_STATUS  encoded_MODALIDAD  encoded_JORNADA  \
0                0               0                  0                0   
1                0               0                  0                0   
2                0               0                  0                0   
3                0               0                  0                0   
4                0               0                  1                1   
...            ...             ...                ...              ...   
9616             0               0                  0                0   
9617             0               0                  0                4   
9618             0               0                  0                3   
9619             0               0                  0                1   
9620             0               0                  0                0   

      encoded_PLAN  
0                0  
1                0  
2                0  
3                0  
4                0  
...            ...  
9616             0  
9617             6  
9618             0  
9619             0  
9620             0  

[8087 rows x 10 columns]

In [21]:
list_of_lists

[array(['PRIVADO', 'OFICIAL', 'MUNICIPAL', 'COOPERATIVA'], dtype=object),
 array(['URBANA', 'RURAL', 'SIN ESPECIFICAR'], dtype=object),
 array(['ABIERTA', 'CERRADA TEMPORALMENTE', 'TEMPORAL NOMBRAMIENTO',
        'TEMPORAL TITULOS'], dtype=object),
 array(['MONOLINGUE', 'BILINGUE'], dtype=object),
 array(['MATUTINA', 'VESPERTINA', 'DOBLE', 'NOCTURNA', 'SIN JORNADA',
        'INTERMEDIA'], dtype=object),
 array(['DIARIO(REGULAR)', 'FIN DE SEMANA', 'A DISTANCIA',
        'SEMIPRESENCIAL', 'SEMIPRESENCIAL (FIN DE SEMANA)', 'SABATINO',
        'SEMIPRESENCIAL (UN DÍA A LA SEMANA)',
        'SEMIPRESENCIAL (DOS DÍAS A LA SEMANA)', 'DOMINICAL',
        'VIRTUAL A DISTANCIA', 'MIXTO', 'INTERCALADO', 'IRREGULAR'],
       dtype=object)]

In [22]:
# Find columns with null values
null_columns = df.isnull().sum()

# Display the result
print(null_columns)

DISTRITO             0
DEPARTAMENTO         0
MUNICIPIO            0
ESTABLECIMIENTO      0
encoded_SECTOR       0
encoded_AREA         0
encoded_STATUS       0
encoded_MODALIDAD    0
encoded_JORNADA      0
encoded_PLAN         0
dtype: int64


Podemos apreciar que el dataframe ya no tiene valores nulos y las dimensiones del mismo se ven reducidas a 8087 filas y 13 columnas.

In [23]:
df.shape

(8087, 10)